io-uring-test/src/tests/register_sync_cancel.rs

```diff
@@ -0,0 +1,105 @@
+    const USER_DATA_0: u64 = 42u64;
+    // 2 operations, canceled by the user_data
+    const USER_DATA_1: u64 = 43u64;
+    // 2 operations, canceled by the fd.
+    const USER_DATA_2: u64 = 44u64;
+
+    // Start reads for user_data_0
+    let mut buf = [0u8; 32];
+    for i in 0..5 {
+        let mut entry = opcode::Read::new(types::Fd(0), buf.as_mut_ptr(), 32).build();
+
+        if i == 0 {
+            entry = entry.user_data(USER_DATA_0);
+        } else if (1..3).contains(&i) {
+            entry = entry.user_data(USER_DATA_1);
+        } else if (3..5).contains(&i) {
+            entry = entry.user_data(USER_DATA_2);
+        }
+        unsafe { ring.submission().push(&entry.into()).unwrap() };
+    }
+    // Submit all 5 operations.
+    assert_eq!(5, ring.submit()?);
+
+    // Cancel the first operation by user_data
+    ring.submitter()
+        .register_sync_cancel(USER_DATA_0, -1, None, AsyncCancelFlags::empty())?;
+    ring.submitter().submit_and_wait(1)?;
+    let cqe: cqueue::Entry = ring.completion().next().unwrap().into();
+    assert_eq!(
+        cqe.result(),
+        -libc::ECANCELED,
+        "operation should have been canceled"
+    );
+    assert_eq!(cqe.user_data(), USER_DATA_0);
+
+    // Cancel the second two operations by their user data flags.
+    ring.submitter()
+        .register_sync_cancel(USER_DATA_1, -1, None, AsyncCancelFlags::ALL)?;
+    ring.submitter().submit_and_wait(2)?;
+    let cqe1: cqueue::Entry = ring.completion().next().unwrap().into();
+    let cqe2: cqueue::Entry = ring.completion().next().unwrap().into();
+    assert_eq!(cqe1.result(), cqe2.result());
+    assert_eq!(cqe1.result(), -libc::ECANCELED);
+
+    // Cancel the last two operations by their fd.
+    ring.submitter().register_sync_cancel(
+        USER_DATA_2,
+        0,
+        None,
+        AsyncCancelFlags::FD | AsyncCancelFlags::ALL,
+    )?;
+    ring.submitter().submit_and_wait(2)?;
+    let cqe1: cqueue::Entry = ring.completion().next().unwrap().into();
+    let cqe2: cqueue::Entry = ring.completion().next().unwrap().into();
+    assert_eq!(cqe1.result(), cqe2.result());
+    assert_eq!(cqe1.result(), -libc::ECANCELED);
+
+    Ok(())
+}
+
+pub fn test_register_sync_cancel_unsubmitted<S: squeue::EntryMarker, C: cqueue::EntryMarker>(
+    ring: &mut IoUring<S, C>,
+    test: &Test,
+) -> io::Result<()> {
+    require!(
+        test; // We need at least 6.0 to use `IORING_REGISTER_SYNC_CANCEL`. opcode::SendZc is a proxy for that requirement.
+        test.probe.is_supported(opcode::SendZc::CODE);
+    );
+    // Test that we can cancel operations which have not yet been submitted.
+    const USER_DATA: u64 = 42u64;
+
+    let mut buf = [0u8; 32];
+    let entry = opcode::Read::new(types::Fd(0), buf.as_mut_ptr(), 32).build();
+    unsafe { ring.submission().push(&entry.into()).unwrap() };
+
+    // Cancel the operation by user_data, we haven't submitted anything yet.
+    ring.submitter()
```

Comment:
This should be failing because we have't submitted the operation, and sync cancel will only work for operations which have been submitted.

Reply:
That's odd. Is it being run in parallel with the other test on the same ring? Have you tried with a different USER_DATA value?

Reply:
Sorry, I meant that this test should be checking for failure rather than success. It was failing previously, but it seems that CI might not be running these tests so it wasn't picked up until I ran it locally.

Reply:
Well that makes sense. I was wondering how that was going to be worked out. I didn't even think that I might have been misunderstanding what was meant.